Feat/issue 4/rate limiter/response model

taralamia opened the pull request:
# 🚀 Pull Request: Implement Content Negotiation, Rate Limiting, and Response Model Middleware

## ✨ Summary
This PR introduces three key middleware functionalities to the backend:
1. **Content Negotiation**
2. **Rate Limiting**
3. **Response Model Standardization**

---

## 🔧 Implementation Details

### **1. Content Negotiation**
Implemented a robust and reusable content negotiation workflow that dynamically serves responses in **JSON**, **XML**, or **HTML** formats based on the client's `Accept` header.

- Added the negotiation logic in the **presentation layer** (`presentation/http/negotiation.ts`).
- Implemented a dedicated **middleware** (`middleware/negotiationMiddleware.ts`) that sends the final negotiated response.
- Added safe handling for:
  - Unsupported formats (`406 Not Acceptable`)
  - Serialization errors (invalid JSON/XML/HTML)
- Added proper handling of the `Vary` header to ensure accurate caching behavior.

### **2. Rate Limiter**
Implemented a rate limiter middleware to protect API endpoints from excessive requests.

- Researched multiple rate-limiting algorithms (Token Bucket, Fixed Window, Sliding Window) from this [blog post](https://medium.com/@ankitjaat24u/implementing-a-rate-limiter-and-slow-down-mechanism-in-next-js-with-typescript-2700d4d6fe0c).
- Chose and integrated an appropriate algorithm suited for the current system needs.
- Built the structure to allow future extensions (e.g., slowdown mechanisms, per-route config).

### **3. Response Model**
Introduced a standardized response model to maintain consistent response structures across all endpoints.

- Every response now follows a unified schema:
  ```json
  {
    "success": true,
    "data": {},
    "error": null
  }
  ```
- Simplifies client-side integration and testing.

---

## ✅ Why These Changes Matter
- **Consistency:** All routes now follow a standard response format.
- **Resilience:** The rate limiter prevents abuse and protects the API under load.
- **Flexibility:** The content negotiation layer improves compatibility with various client applications.

taralamia committed:
feat(middleware): Add centralized API rate limiting with AppError integration #4

taralamia committed:
Implement rate limiting and fix response delay by introducing response negotiation wrapper
- Added custom withNegotiation wrapper to ensure consistent and timely API responses across controllers.
- Resolved Postman delay issue caused by missing response handling middleware.
- Configured API rate limiting with a short time window and limited request count for testing. #4

parvez-ahammed commented on src/routers/authRoutes.ts:
```diff
@@ -23,10 +23,8 @@ const authService = new AuthService(
-router.post("/auth/signin",authController.signIn);
-router.post("/auth/refresh",authController.refresh);
-router.use(negotiationMiddleware);
+router.post("/auth/signup",requireJsonBody(),withNegotiation(authController.signUp));
```
Instead of setting up negotiations in every route. We can consider this as an global middleware. So we can think like this that, at the start of the request it check if the content it is seeking is acceptable or not. Then at the while returning the request it just formats it in that way. As currently we have a respone model we should be easily able to parse the final model. This will reduce the redundance.

taralamia replied:
Noted and I've applied your idea , kindly checkout the recent commits to evaluate my work.

parvez-ahammed commented on src/server.ts:
```diff
@@ -3,11 +3,13 @@ import { AppDataSource } from "./data-source";
 const PORT = env.PORT;
 app.use(express.json());
 app.use(cookieParser());
+app.use(rateLimiter);
```
We can use the content negotiations middleware in this level.

parvez-ahammed reviewed (approved):
Good work so far. Amazing PR descriptions. Approved the PR with small comments.

taralamia committed:
refactor: apply content negotiation via sendNegotiated and remove finalize middleware

- Removed preNegotiationMiddleware() and finalizeResponse() to simplify request flow
- Integrated sendNegotiated() in VehicleController (list, get, update, remove) to support XML/JSON/HTML responses
- Updated AuthController methods (signUp, verifyEmail, signIn, refresh) to use sendNegotiated() for content negotiation
- Ensured cookies, headers, and cache controls are set before negotiation
- Maintained consistent response structure across controllers
- Verified responses handling via Accept header and adjusted SUPPORTED_TYPES accordingly #4

taralamia committed:
chore: Remove the unused files #4

parvez-ahammed commented:
Looks good to me. Let's merge it.
